src/entities/amm.ts

```diff
@@ -258,31 +338,538 @@ class AMM {
+
+    const { closestUsableTick: tickUpper } = this.closestTickAndFixedRate(fixedLow);
+    const { closestUsableTick: tickLower } = this.closestTickAndFixedRate(fixedHigh);
+
+    let sqrtPriceLimitX96;
+    if (fixedRateLimit) {
+      const { closestUsableTick: tickLimit } = this.closestTickAndFixedRate(fixedRateLimit);
+      sqrtPriceLimitX96 = TickMath.getSqrtRatioAtTick(tickLimit).toString();
+    } else {
+      if (isFT) {
+        sqrtPriceLimitX96 = TickMath.getSqrtRatioAtTick(TickMath.MAX_TICK - 1).toString();
+      } else {
+        sqrtPriceLimitX96 = TickMath.getSqrtRatioAtTick(TickMath.MIN_TICK + 1).toString();
+      }
+    }
+
+    const peripheryContract = peripheryFactory.connect(this.peripheryAddress, this.signer);
+    const scaledNotional = this.scale(notional);
+
+    let swapPeripheryParams: SwapPeripheryParams;
+    let tempOverrides: {value?: BigNumber, gasLimit?: BigNumber} = {};
+    
+
+    if (this.isETH) {
+      swapPeripheryParams = {
+        marginEngine: newMarginEngine,
+        isFT,
+        notional: scaledNotional,
+        sqrtPriceLimitX96,
+        tickLower,
+        tickUpper,
+        marginDelta: '0',
+      };
+
+      tempOverrides.value = ethers.utils.parseEther(margin.toString());
+    }
+    else {
+      const scaledMarginDelta = this.scale(margin);
+
+      swapPeripheryParams = {
+        marginEngine: newMarginEngine,
+        isFT,
+        notional: scaledNotional,
+        sqrtPriceLimitX96,
+        tickLower,
+        tickUpper,
+        marginDelta: scaledMarginDelta,
+      };
+    }
+
+    await peripheryContract.callStatic.rolloverWithSwap(
+      this.marginEngineAddress,
+      effectiveOwner,
+      oldTickLower,
+      oldTickUpper,
+      swapPeripheryParams,
+      tempOverrides
+    ).catch(async (error: any) => {
+      const errorMessage = getReadableErrorMessage(error, this.environment);
+      throw new Error(errorMessage);
+    })
+
+    const estimatedGas = await peripheryContract.estimateGas.rolloverWithSwap(
+      this.marginEngineAddress,
+      effectiveOwner,
+      oldTickLower,
+      oldTickUpper,
+      swapPeripheryParams,
+      tempOverrides
+    ).catch((error) => {
+      const errorMessage = getReadableErrorMessage(error, this.environment);
+      throw new Error(errorMessage);
+    });
+
+    tempOverrides.gasLimit = getGasBuffer(estimatedGas);
+
+    const swapTransaction = await peripheryContract.rolloverWithSwap(
+      this.marginEngineAddress,
+      effectiveOwner,
+      oldTickLower,
+      oldTickUpper,
+      swapPeripheryParams,
+      tempOverrides
+    ).catch((error) => {
+      const errorMessage = getReadableErrorMessage(error, this.environment);
+      throw new Error(errorMessage);
+    });
+
+    try {
+      const receipt = await swapTransaction.wait();
+      return receipt;
+    }
+    catch (error) {
+      throw new Error("Transaction Confirmation Error");
+    }
+  }
+
+  public async getInfoPostRolloverWithSwap({
```

Comment:
good for now but it'd be better for us to restructure this code in order to avoid this code duplicate in here.